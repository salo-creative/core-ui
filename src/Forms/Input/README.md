# Input

This is an input component. It works well with `useForm`.

## Usage

Install

```bash
yarn add @salo/core-ui
```

Include

```javascript
import Input from '@salo/core-ui/Forms/Input';
```

Implement as follows

```javascript
<Input 
  name='email'
  placeholder='Enter your email address'
/>
```

## Icons

There are a couple of ways you can pass icons to the input.

1. The simplest way is with a string, this will default to left align with a left position of 3rem.

```javascript
<Input icon='pencil' />
```

2. If you need granular control of the icon you can pass an object like this:

```javascript
<Input 
  icon={ {  
    fill: PropTypes.string,
    icon: PropTypes.string.isRequired,
    offset: PropTypes.string,
    position: PropTypes.oneOf(['left', 'right'])
  } }  
/>
```

This should give you full flexibility to position and style your icon.