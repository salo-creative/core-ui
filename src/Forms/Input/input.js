import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS
import { InputWrapper, Field, FieldWrapper } from './input.styles';
import Label from '../components/Label';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import Strength from '../components/Strength';

// HELPERS
import { sanitize } from '../../helpers/form';
import { colours } from '../../helpers/colours';

const Input = (props) => {
  const {
    background,
    className,
    disabled,
    error,
    errorMessage,
    fontSize,
    helperText,
    label,
    margin,
    name,
    onBlur,
    onChange,
    onKeyUp,
    padding,
    placeholder,
    required,
    showStrength,
    size,
    type,
    value,
    border,
    borderRadius
  } = props;

  const height = () => {
    switch (size) {
      case 'M':
        return '4rem';
      case 'L':
      default:
        return '4.5rem';
    }
  };

  return (
    <InputWrapper margin={ margin } className={ `salo-input ${ className }` }>
      <Label
        error={ error }
        label={ label }
        name={ name }
        required={ required }
        size={ size }
        className='salo-input__label'
      />
      <FieldWrapper className='salo-input__wrapper'>
        <Field
          aria-required={ required.toString() }
          aria-invalid={ error.toString() }
          background={ background }
          border={ border }
          borderRadius={ borderRadius }
          className='salo-input__field'
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          height={ height() }
          id={ name }
          name={ name }
          onBlur={ (e) => onBlur({ e, value: sanitize(e.target.value) }) }
          onChange={ (e) => onChange({ e, value: sanitize(e.target.value) }) }
          onKeyUp={ (e) => onKeyUp({ e, value: sanitize(e.target.value) }) }
          padding={ padding }
          placeholder={ placeholder }
          type={ type }
          value={ value }
        />
      </FieldWrapper>
      <ErrorText
        className='salo-input__error'
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        className='salo-input__helper'
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
      { type === 'password' && showStrength && (<Strength value={ value } className='salo-input__strength' />) }
    </InputWrapper>
  );
};

Input.defaultProps = {
  background: colours.paleGrey,
  border: '1px solid',
  borderRadius: '0.4rem',
  className: '',
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  fontSize: '1.4rem',
  helperText: '',
  label: '',
  margin: '0 0 2rem',
  onBlur: () => null,
  onChange: () => null,
  onKeyUp: () => null,
  padding: '0 1rem',
  placeholder: '',
  required: false,
  showStrength: false,
  size: 'M',
  type: 'text',
  value: ''
};

Input.propTypes = {
  background: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  fontSize: PropTypes.string,
  helperText: PropTypes.string,
  label: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onKeyUp: PropTypes.func,
  padding: PropTypes.string,
  placeholder: PropTypes.string,
  border: PropTypes.string,
  borderRadius: PropTypes.string,
  required: PropTypes.bool,
  showStrength: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  type: PropTypes.oneOf(['text', 'email', 'tel', 'number', 'password', 'chat']),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default Input;