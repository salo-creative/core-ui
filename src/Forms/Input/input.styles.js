import styled from 'styled-components';
import { transparentize } from 'polished';

export const InputWrapper = styled.div`
  width: 100%;
  flex-direction: column;
  display: flex;
  margin: ${ ({ margin }) => margin };
`;

export const FieldWrapper = styled.div`
  position: relative;
`;

export const Field = styled.input`
  margin: 0;
  padding: ${ ({ padding }) => padding };
  height: ${ ({ height }) => height };
  background-color: ${ ({ background, error, theme }) => (error ? transparentize(0.7, theme.error) : background) };
  border: ${ ({ border }) => border };
  border-color: ${ ({ theme, error }) => (error ? transparentize(0.3, theme.error) : transparentize(0.5, theme.blue)) };
  transition: border 0.2s linear;
  width: 100%;
  border-radius: ${ ({ borderRadius }) => borderRadius };
  font-size: ${ ({ fontSize }) => fontSize };

  &:focus,
  &:active {
    border-color: ${ ({ theme }) => theme.primary }
  }

  &[disabled] {
    cursor:  not-allowed;
    opacity: 0.75;
  }
`;