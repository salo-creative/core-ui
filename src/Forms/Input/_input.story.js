import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Input } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Forms | Input', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['input'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: 'Some text' });
    // KNOBS
    const label = text('Label', 'Input label');
    const helperText = text('Helper text', 'Helper text');
    const icon = text('Icon', 'pencil');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const showStrength = boolean('Show password strength indicator', false);
    const size = select('size', ['L', 'M'], 'M');
    const type = select('type', ['text', 'email', 'tel', 'number', 'password'], 'text');
    return (
      <RenderWithProps store={ store }>
        <Input
          name='story'
          disabled={ disabled }
          error={ error }
          label={ label }
          required={ required }
          helperText={ helperText }
          showStrength={ showStrength }
          size={ size }
          type={ type }
          icon={ icon }
          value={ store.value }
          onChange={ ({ value }) => store.set({ value }) }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);