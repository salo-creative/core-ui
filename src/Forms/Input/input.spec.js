import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Input from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Input />
  );
  expect(container.firstChild).toMatchSnapshot();
});