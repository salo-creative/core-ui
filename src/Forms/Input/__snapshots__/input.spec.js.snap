// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem;
}

.c1 {
  position: relative;
}

.c2 {
  margin: 0;
  padding: 0 1rem;
  height: 4rem;
  background-color: #fafcff;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  width: 100%;
  border-radius: 0.4rem;
  font-size: 1.4rem;
}

.c2:focus,
.c2:active {
  border-color: #00ACED;
}

.c2[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

<div
  class="salo-input  c0"
>
  <div
    class="salo-input__wrapper c1"
  >
    <input
      aria-invalid="false"
      aria-required="false"
      class="salo-input__field c2"
      font-size="1.4rem"
      height="4rem"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;
